python/ray/serve/long_poll.py

```diff
@@ -0,0 +1,196 @@
+import asyncio
+from asyncio.tasks import Task
+import random
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any, Awaitable, Callable, DefaultDict, Dict, Generic, Set, Tuple, List, TypeVar, Optional
+
+import ray
+from ray.serve.utils import logger
+
+
+@dataclass
+class UpdatedObject:
+    object_snapshot: Any
+    # The identifier for the object's version. There is not sequential relation
+    # among different object's snapshot_ids.
+    snapshot_id: int
+
+
+# Type signature for the update state callbacks. E.g.
+# def update_state(updated_objects: Dict[str, Any], updated_keys: List[str]):
+#     if "my_key" in updated_keys: do_something(updated_objects["my_key"])
+UpdateStateCallable = Callable[[Dict[str, Any], List[str]], None]
+UpdateStateAsyncCallable = Callable[[Dict[str, Any], List[str]], Awaitable[
+    None]]
+
+
+class BaseClient:
+    """Shared functionality between LongPollerSyncClient and AsyncClient.
+       This class is not expected to use this directly.
+    """
+
+    def __init__(self, host_actor, keys: List[str]) -> None:
+        self.host_actor = host_actor
+        self.snapshot_ids: Dict[str, int] = {key: -1 for key in keys}
+        self.object_snapshots: Dict[str, Any] = dict()
+
+    def _pull_once(self) -> ray.ObjectRef:
+        object_ref = self.host_actor.listen_on_changed.remote(
+            self.snapshot_ids)
+        return object_ref
+
+    def _update(self, updates: Dict[str, UpdatedObject]):
+        for key, update in updates.items():
+            self.object_snapshots[key] = update.object_snapshot
+            self.snapshot_ids[key] = update.snapshot_id
+
+    def get_object_snapshot(self, object_key: str) -> Any:
+        return self.object_snapshots[object_key]
+
+
+class LongPollerSyncClient(BaseClient):
+    """The synchronous long polling client.
+
+    Internally, it uses ray.wait(timeout=0) to check if the object notification
+    has arrived. When a object notification arrived, the client will invoke
+    callback if supplied.
+    """
+
+    def __init__(self,
+                 host_actor,
+                 keys: List[str],
+                 callback: Optional[UpdateStateCallable] = None) -> None:
+        super().__init__(host_actor, keys)
+        self.in_flight_request_ref: ray.ObjectRef = self._pull_once()
+        self.callback = callback
```

Comment:
Given that this is called synchronously anyways, does it make sense to have a callback? I think the simplest interface would be to have just `refresh` and it returns the updated keys. Then the caller can call their function however they want to. We can get rid of the get_object_snapshot method too in that case.